src/hugr/validate.rs

```diff
@@ -63,6 +71,33 @@ impl<'a> ValidationContext<'a> {
         Ok(())
     }
 
+    /// Use the signature supplied by a dataflow node to work out the resource
+    /// resource requirements for all of its input and output edges, then put
+    /// those requirements in the ValidationContext
+    fn gather_resources(&mut self, node: &Node) -> Result<(), ValidationError> {
+        let op = self.hugr.op_types.get(node.index);
+        let sig = op.signature();
+        let input_ports = self.hugr.node_ports(node.clone(), Direction::Incoming);
+
+        for p in input_ports.into_iter() {
+            let port_ix = self.hugr.graph.port_index(node.index, p.offset).unwrap();
+            assert!(self
+                .resources
+                .insert(port_ix, sig.input_resources.clone())
+                .is_none());
+        }
+
+        let output_ports = self.hugr.node_ports(node.clone(), Direction::Outgoing);
+        for p in output_ports.into_iter() {
+            let port_ix = self.hugr.graph.port_index(node.index, p.offset).unwrap();
+            assert!(self
+                .resources
+                .insert(port_ix, sig.output_resources.clone())
+                .is_none());
+        }
```

Comment:
If you add a `Signature::get_resources(dir: Direction)` (similar to the other methods) this should be doable without repetition
```rust
        for dir in Direction::BOTH {
            for p in self.hugr.node_ports(node.clone(), dir) {
                let port_ix = self.hugr.graph.port_index(node.index, p.offset).unwrap();
                assert!(self
                    .resources
                    .insert(port_ix, sig.get_resources(dir).clone())
                    .is_none());
            }
        }
```